src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/IServiceCollectionExtensions.cs

```diff
@@ -191,6 +192,7 @@ public static void AddDocumentManagementServices(this IServiceCollection service
         services.AddSingleton<GeneratedDocumentPublisher, DefaultGeneratedDocumentPublisher>();
         services.AddSingleton<IProjectSnapshotChangeTrigger>((services) => services.GetRequiredService<GeneratedDocumentPublisher>());
         services.AddSingleton<DocumentContextFactory, DefaultDocumentContextFactory>();
+        services.AddSingleton(sp => new Lazy<DocumentContextFactory>(sp.GetRequiredService<DocumentContextFactory>));
```

Comment:
This is the annoying way to do lazy DI in this system, and its annoying, but I didn't think it was worth extracting new services and things, because ideally we can move to pull diagnostics in VS Code soon, and then the entire diagnostic publisher can be deleted.